### ISSUE TITLE
Uses test data to select models during training

### ISSUE THREAD

**Issue by connellybarnes** [kuangliu/pytorch-cifar#14]
Based on my reading of this code, it appears like there is currently a split into only train and test datasets, and test set accuracy is used to select the best model during training. This appears to result in a quite overoptimistic estimate of the test performance of the model (especially due to the 3 stages of learning with different learning rates), because test accuracies were already used for training time model selection (the model is "cheating" by learning on the test data). In particular, for the VGG16 model I noticed the test accuracy jumps by around +10% between the end of the first training stage and the beginning of the second training stage (i.e. between lr=0.1 and lr=0.01).

In theory, if you do want to select models without this "cheating" effect then a way to address this is to split into 3 subsets of the data: train, test, and validation, use validation to pick the best model, and only ever evaluate your model on the test set, and do not choose between models based on test set accuracy. See this StackExchange post for details:

https://stats.stackexchange.com/questions/152907/how-do-you-use-test-data-set-after-cross-validation

However, in practice it seems that for CIFAR10 the models simply need more regularization by using both batch normalization and dropout, and it suffices to simply always report the test accuracy at the latest epoch (thus in practice, there is no need for the 3 way test/train/validate split). See e.g. the below repo or paper:

https://github.com/chengyangfu/pytorch-vgg-cifar10
http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7486599

**Comment by kuangliu** [comment]
- CIFAR10 is provided as 'train+val'(http://www.cs.toronto.edu/~kriz/cifar.html), no 'test' set. And normally we don't split a 'test' set out, just use 'train+val'. So any papers using CIFAR10, the accuracy is always validation accuracy.
- My code uses the word 'test' to represent 'validation'.
- For dropout, if the paper doesn't use it, you may not find it in the code. My experience, with BN, dropout is no longer necessary.

**Comment by xuanqing94** [comment]
+1, the problem is not reporting val accuracy, but the code uses best model to initialize next stage, rather than the model in last epoch.
